### ISSUE TITLE
Automatically Include Common References in Examples

### ISSUE THREAD

**Issue by Ant13731** [JacquesCarette/Drasil#2675]
Building off of @JacquesCarette's comments in #2279,

> Note that, over time, we want to automate these parts. These lists of documentation and adding them to the right places in the examples should not be a step that a human needs to do. In fact, for `doccon` and `doccon'`, that might be possible now.

We should be able to automate some references so that users do not have to manually place them in the `ChunkDB`. It might be too early to automate references created in specific examples, but references that are more universal (like `doccon`,`doccon'`, and those references listed in `DocLang/SRS/References.hs`) should be automatically placed in the `ChunkDB` for every example.

Completing this issue will also require modifying the instructions for [Creating Your Project in Drasil](https://github.com/JacquesCarette/Drasil/wiki/Creating-Your-Project-in-Drasil#adding-introduction-section).

**Comment by Ant13731** [comment]
For the `IdeaDict` map in the chunk database (where things like `doccon` and `doccon'` are placed), should I include all of the concepts across all the domains we currently have?

 For example, should I include `softwarecon`, `educon`, `mathcon`, `compcon`, `physicCon`, `solidcon`, and `thermocon`? Everything except the last three almost always appear in the examples, so I think those should always be included. But should I automatically include `physicCon`, `solidcon` and `thermocon` the database, or let the user add them when they are needed?

**Comment by JacquesCarette** [comment]
What we really should be doing is the opposite: only include the ones that are actually used. Automatically, not by hand.

All the currently hand-built lists of chunks in the examples should, over time, disappear. We don't want to fake this by moving the lists to other parts of Drasil either.

**Comment by Ant13731** [comment]
Okay, I think I see what you mean, but I'm not too sure how to go about it fully automatically. 

My first thought was to have a second database running in the background (so that the user does not know about it) with all of the chunks from `drasil-data`. Then the example would check the `UIDs` used in the example and pull from the "background" database into its actual database. But this just feels like moving around the list of chunks rather than automatically including them when needed. It could also lead to unexpected behaviour if the same UID is used in `drasil-data` and the example when the user doesn't know about it.

My next idea would be to somehow tag all the named chunks in `drasil-data` with their respective domains (eg. Just make all the `NamedChunks` into `CIs` so the long list of chunks like `compcon` would all have the `compScience` tag) and have the user include which domains are needed for their project (maybe as a new field in `SystemInformation`), but I think that would still need a "background" database to hold all the chunks from `drasil-data`.

I feel like these two ideas don't really improve the situation, they just hide it better. What do you think I should do?

Also, I think automatically including references is going fairly well, since all of the sections, labelledContent, ConceptInstances, models and definitions are already included in the database for each example, so I can just pull the references from there. Is this more along the lines of what we should aim for?

**Comment by Ant13731** [comment]
Also, while we're on this topic, I noticed a lot of the examples have similar imports from `docLang`, `printers`, and `database`. Would it be worth it to re-export those into another module (we could call it something like Drasil.SRSDocument) so that we wouldn't have so many imports at the beginning of every Body.hs file? For example, the top of SWHS/Body.hs looks like:
``` Haskell
import Data.List (nub)
import Language.Drasil hiding (organization, section, variable)
import Language.Drasil.Printers (PrintingInformation(..), defaultConfiguration, piSys)
import Database.Drasil (Block, ChunkDB, ReferenceDB,
  SystemInformation(SI), cdb, rdb, refdb, _authors, _purpose, _concepts, _constants,
  _constraints, _datadefs, _instModels, _configFiles, _defSequence, _inputs, _kind, 
  _outputs, _quants, _sys, _sysinfodb, _usedinfodb, _folderPath)
import Theory.Drasil (GenDefn, InstanceModel)
import Utils.Drasil
import Utils.Drasil.Concepts
import qualified Utils.Drasil.NounPhrase as NP
import qualified Utils.Drasil.Sentence as S

import Control.Lens ((^.))

import Drasil.DocLang (AuxConstntSec (AuxConsProg), DocSection (..),
  Field(..), Fields, LFunc(TermExcept), Literature(Doc', Lit), IntroSec(IntroProg),
  IntroSub(IChar, IOrgSec, IPurpose, IScope), RefSec (RefProg), 
  RefTab (TAandA, TUnits), TSIntro(SymbConvention, SymbOrder, TSPurpose, VectorUnits),
  ReqrmntSec(..), ReqsSub(..), SRSDecl, SSDSub(..), SolChSpec (SCSProg),
  SSDSec(..), InclUnits(..), DerivationDisplay(..), SCSSub(..), Verbosity(..),
  TraceabilitySec(TraceabilityProg), GSDSec(..), GSDSub(..),
  ProblemDescription(PDProg), PDSub(..), intro, mkDoc, fillTraceSI, tsymb'', traceMatStandard, purpDoc, getTraceConfigUID,
  secRefs, traceyGraphGetRefs)
import qualified Drasil.DocLang.SRS as SRS (inModel)
```
but we could probably reduce it to:
```Haskell
import Data.List (nub)
import Language.Drasil hiding (organization, section, variable)
import Drasil.SRSDocument
import Theory.Drasil (GenDefn, InstanceModel)
import Utils.Drasil
import Utils.Drasil.Concepts
import qualified Utils.Drasil.NounPhrase as NP
import qualified Utils.Drasil.Sentence as S

import Control.Lens ((^.))
import qualified Drasil.DocLang.SRS as SRS (inModel)
```
where
```Haskell
module Drasil.SRSDocument where

import Drasil.DocLang (AuxConstntSec (AuxConsProg), DocSection (..),
  Field(..), Fields, LFunc(TermExcept), Literature(Doc', Lit), IntroSec(IntroProg),
  IntroSub(IChar, IOrgSec, IPurpose, IScope), RefSec (RefProg), 
  RefTab (TAandA, TUnits), TSIntro(SymbConvention, SymbOrder, TSPurpose, VectorUnits),
  ReqrmntSec(..), ReqsSub(..), SRSDecl, SSDSub(..), SolChSpec (SCSProg),
  SSDSec(..), InclUnits(..), DerivationDisplay(..), SCSSub(..), Verbosity(..),
  TraceabilitySec(TraceabilityProg), GSDSec(..), GSDSub(..),
  ProblemDescription(PDProg), PDSub(..), intro, mkDoc, fillTraceSI, tsymb'', traceMatStandard, purpDoc, getTraceConfigUID,
  secRefs, traceyGraphGetRefs)

import Language.Drasil.Printers (PrintingInformation(..), defaultConfiguration, piSys)
import Database.Drasil (Block, ChunkDB, ReferenceDB,
  SystemInformation(SI), cdb, rdb, refdb, _authors, _purpose, _concepts, _constants,
  _constraints, _datadefs, _instModels, _configFiles, _defSequence, _inputs, _kind, 
  _outputs, _quants, _sys, _sysinfodb, _usedinfodb, _folderPath)
```
This way, users only get the imports they need, but we only have to change the imports in one place rather than every example (when we need to change something across all examples).

**Comment by Ant13731** [comment]
I've also been thinking on the original issue some more. If we have a secondary "background" database, then we could traverse through the whole `DocDesc` to find which terms we need (through pattern matching on `Ch`) and then transfer the needed UIDs over to our `SystemInformation` database. I think the `Multiplate DLPlate` made in `DocumentLanguage/Core.hs` would be able to help, but I'm not too sure how to use it (or where to get started in learning how to use it).

**Comment by JacquesCarette** [comment]
Having an easier set of imports for the examples is a good idea (and a separate issue). Go ahead on that.

**Comment by JacquesCarette** [comment]
As to the main topic: yes, there is going to be a need for a "background database". What I mean is that people who make examples should have very little need to put things into ` ChunkDB` "by hand". The act of using them in an SRS should be (close to) enough.

But that can't quite work since the SRS often contains only UIDs. So when doing an example, the 'knowledge base' that is being built up specifically for the example still needs to be created. So some kind of "registration process" of the information into the knowledge base is still needed for those things that only even appear as UIDs.

In other words, this is going to be an incremental process, i.e. this automation will occur over time. The more the chunks themselves appear in the SRS (instead of just their UIDs), the more we can automate.

**Comment by Ant13731** [comment]
Is the end goal to not need the database to generate the document in the first place (i.e. no registration process)? As in, the document is self sufficient and the database is like an artifact created from making the SRS? Would that entail including actual chunks inside of sentences rather than just the chunk UIDs? So something like the `Ch` constructor (which takes in a `UID` to make a Sentence) should hold the chunk itself? If that's the case, was making the `Ref` constructor only able to hold `UIDs` rather than actual `References` moving in the wrong direction? In the end, would #1541 no longer be relevant since the database would just be created from the SRS?

Or is it more along the lines of using some parts of the SRS to generate the "background database" (that will then be used in the rest of the SRS). Something where users put information into the SRS (through datadefs, gendefs, etc.), and from there Drasil builds the chunk database that includes all of the terms, symbols, and other information so that it can be used in Sentences, the table of units, table of symbols, and table of acronyms?

**Comment by Ant13731** [comment]
Also, would this mean that types like `ConceptChunks` should eventually hold a list of domains as chunks rather than just the domain's UIDs?

**Comment by JacquesCarette** [comment]
The end goal is to make the writing of examples simpler, with less overhead on the part of the writer.

So we can still have a database, but maybe if 'registration' was mostly implicit. Having chunks in the actual sentences would help, but that doesn't really work so well with the current setup - we'd have to teach `Sentence` about our chunks, which is not something I really want to do.

**Comment by Ant13731** [comment]
Okay, I think I see what you mean by an incremental process. It's not really feasible to create a database without any user overhead all at once.  So how would we go about implementing an implicit registration? It would be nice if a user could just use the chunk once and have it registered. But they would still need to import the needed chunks into a file and as of now still import them into one or more data bases. So would a good first step be getting rid of the second database (`usedDB`)? I think there's an issue about it somewhere and its involvement in the table of units, symbols, and abbreviations.

**Comment by Ant13731** [comment]
Closing this in favour of #2775, since the original issue has been completed.
